### ISSUE TITLE
Slow query response times

### ISSUE THREAD

**Issue by geekhead** [jacksontj/promxy#188]
### Issue 
A simple query using the `up{}` metric for the last 7 days or even just the last 3-4 results in a 3-5 second delay. The same query going directly to either of the VictoriaMetrics backends results in only a 130-170ms delay. I understand that Promxy has to query all the servers in the server group and then merge the results but a ~180% increase in query times seems a little much for my simple setup.  

### Setup
So I have a very simple setup of two instances. Each instance is running VictoriaMetrics (version 1.20.7) and Prometheus (2.9.2). I have one of those instances running Promxy (v0.0.41) with the two VictoriaMetrics instances in a server group. 

Promxy config:
```
global:
  evaluation_interval: 5s
  external_labels:
    source: promxy

promxy:
  server_groups:
    - static_configs:
      - targets:
        - prometheus1-a:8428
        - prometheus1-b:8428
      labels:
        sg: VM_Cluster1
      anti_affinity: 10s
      remiote_read: false
      http_client:
        diial_timeout: 2s
        tls_config:
          insecure_skip_verify: true
```

### Results
Going directly to a VictoriaMetric instance
![Screen Shot 2019-07-02 at 11 56 28 AM](https://user-images.githubusercontent.com/2513533/60528496-8c1fe600-9cc2-11e9-868c-1a067df9ef95.png)

![Screen Shot 2019-07-02 at 11 56 38 AM](https://user-images.githubusercontent.com/2513533/60528495-8c1fe600-9cc2-11e9-9a0b-0f37b9696641.png)

Going directly to Promxy
![Screen Shot 2019-07-02 at 11 54 47 AM](https://user-images.githubusercontent.com/2513533/60528683-f6d12180-9cc2-11e9-9d5a-66738375b7e0.png)

![Screen Shot 2019-07-02 at 11 55 01 AM](https://user-images.githubusercontent.com/2513533/60528682-f6d12180-9cc2-11e9-977b-d3e7ba419c2d.png)

Here's the actual Promxy log line entries when doing the following queries:
**Last 2 days:** `[02/Jul/2019 16:22:00] "GET /api/v1/query_range HTTP/1.1 200 6945" 1.794222 start=1561911720&end=1562084520&step=120&query=up%7B%7D`
**Last 7 days:**`[02/Jul/2019 16:23:07] "GET /api/v1/query_range HTTP/1.1 200 5163" 3.456831 start=1561479600&end=1562084400&step=300&query=up%7B%7D`

Let me know if you need any other additional info. Thanks

**Comment by geekhead** [comment]
For more background info, I have prometheus scrapping targets every second so that I can ingest as many data points as I can to also test VictoriaMetrics and it's stability and data storage characteristics. So as of right now, each VictoriaMetrics instance has about 2.7 billion data points which might be compounding the issue with how Promxy determines what to merge into a single global view potentially.

**Comment by jacksontj** [comment]
Thanks for the report, that is more of a slowdown then I'd expect :)

TLDR in this case promxy will do the same `query_range` to all the downstreams then merge the data. The "merge" time should be minimal (definitely way less than a second), so this is interesting. I've added better debug logging for timing (#189 -- merged into master), could you do the same query with debug logging on and paste that output?

Also, when looking at the screenshots of the xhr requests; I see that the response from promxy is 1/10th the size -- is it still correct? Or is it getting an error of some type (or missing data)?

**Comment by geekhead** [comment]
@jacksontj I tried doing a `make release` with the master branch but get the following errors:
```
# github.com/jacksontj/promxy/vendor/github.com/prometheus/prometheus/scrape
/root/go/src/github.com/jacksontj/promxy/vendor/github.com/prometheus/prometheus/scrape/scrape.go:276:11: sp.client.CloseIdleConnections undefined (type *http.Client has no field or method CloseIdleConnections)
/root/go/src/github.com/jacksontj/promxy/vendor/github.com/prometheus/prometheus/scrape/scrape.go:334:11: oldClient.CloseIdleConnections undefined (type *http.Client has no field or method CloseIdleConnections)
An error has occurred! Aborting the script execution...
make: *** [release] Error 1
```

**Comment by jacksontj** [comment]
@geekhead ah, you are likely building with an older version of go, that option was added in go 1.12 :)

**Comment by geekhead** [comment]
And that would be accurate :) `go version go1.11.3 linux/amd64`. I'll get that updated, thanks.

**Comment by geekhead** [comment]
Ok here are the debug logs for doing a ~5 day query:
```
DEBU[2019-07-03T21:55:16Z] http://prometheus1-b:8428                     api=GetValue end="2019-07-03 21:50:31 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-07-03 21:45:31 +0000 UTC"
DEBU[2019-07-03T21:55:16Z] http://prometheus1-a:8428                     api=GetValue end="2019-07-03 21:50:31 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-07-03 21:45:31 +0000 UTC"
DEBU[2019-07-03T21:55:16Z] http://prometheus1-a:8428                     api=GetValue end="2019-07-03 21:50:31 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-07-03 21:45:31 +0000 UTC" took=3.968392ms
DEBU[2019-07-03T21:55:16Z] http://prometheus1-b:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC"
DEBU[2019-07-03T21:55:16Z] http://prometheus1-a:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC"
DEBU[2019-07-03T21:55:16Z] http://prometheus1-b:8428                     api=GetValue end="2019-07-03 21:50:31 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-07-03 21:45:31 +0000 UTC" took=9.98202ms
DEBU[2019-07-03T21:55:16Z] Select                                        matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" selectParams="&{1562190331000 1562190631000 0 }" took=10.285411ms
172.31.69.141 - - [03/Jul/2019 21:55:16] "GET /api/v1/query HTTP/1.1 200 307" 0.011422 query=netdata_system_cpu_percentage_average%7Bdimension%3D%22user%22%7D&time=1562190631
172.31.65.31 - - [03/Jul/2019 21:55:16] "GET /graph HTTP/1.1 200 4850" 0.001364
DEBU[2019-07-03T21:55:19Z] http://prometheus1-b:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC" took=3.291869264s
DEBU[2019-07-03T21:55:19Z] http://prometheus1-a:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC" took=3.372024412s
DEBU[2019-07-03T21:55:19Z] Select                                        matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" selectParams="&{1561751700000 1562190600000 600000 }" took=3.399487815s
172.31.69.141 - - [03/Jul/2019 21:55:19] "GET /api/v1/query_range HTTP/1.1 200 20476" 3.456688 query=netdata_system_cpu_percentage_average%7Bdimension%3D%22user%22%7D&start=1561752000&end=1562190600&step=600
```

**Comment by jacksontj** [comment]
@geekhead Thanks for the output -- and you have uncovered a perf bug :) I'll explain

From the logs you provided it seems that the vast majority of time is spent waiting on the downstream to respond, for example:

```
DEBU[2019-07-03T21:55:19Z] http://prometheus1-b:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC" took=3.291869264s
DEBU[2019-07-03T21:55:19Z] http://prometheus1-a:8428                     api=GetValue end="2019-07-03 21:50:00 +0000 UTC" matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" start="2019-06-28 19:55:00 +0000 UTC" took=3.372024412s
DEBU[2019-07-03T21:55:19Z] Select                                        matchers="[dimension=\"user\" __name__=\"netdata_system_cpu_percentage_average\"]" selectParams="&{1561751700000 1562190600000 600000 }" took=3.399487815s
```

Thats 3.291869264s and 3.372024412s for the 2 downstreams (done in parallel) which ends up being 3.399487815s total time -- which is ~20ms overhead.

So that would indicate that there is no issue? Wrong :) The issue is the query type that is being sent, if you look its doing a full `GetValue` (which gets all the raw datapoints) -- which is not at all required to fulfill the query -- it should be doing a query or query_range instead. From my memory (and the comments) I thought I had already done so -- but apparently not :)

This is actually a trivial fix which I have outstanding in #192 if you give that a shot you should see *significant* performance improvements for cases where the query sent to the downstreams is a VectorSelector.

**Comment by jacksontj** [comment]
@geekhead Sorry about the autoclose, I forget about that ;) I'll leave this open to make sure it solves your issue -- but I'm pretty confident it will :)

**Comment by jacksontj** [comment]
https://github.com/jacksontj/promxy/releases/tag/v0.0.42 has the fix (if you'd rather just pull a binary :) )

**Comment by geekhead** [comment]
@jacksontj just tested it using the latest version and huge improvement. It also fixed the OOM issue when pulling 7 days worth of data so thanks!

**Comment by jacksontj** [comment]
Glad to hear :) Definitely let me know if you find anything else :smile:
